### ISSUE TITLE
Tensorflow 2.1: How does the metric 'tf.keras.metrics.PrecisionAtRecall' works with multiclass-classification? 

### ISSUE THREAD

**Issue by FrancescaAlf** [tensorflow/tensorflow#37256]
I am wondering how this metrics works in case of multiclass classification. Does it compute the average between values of precision belonging to each class?

I have a multiclass-classification problem, with three classes. I am interested in calculate the PrecisionAtRecall when the recall value is equal to 0.76, only for a specific class . What I have done was just setting 0.76 in brackets :
'' model.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy', tf.keras.metrics.PrecisionAtRecall(0.76)], sample_weight_mode='temporal') ''
But I am sure is not the right way.
How can I adjust the metric to reach my goal?
Thanks

**Comment by gowthamkpr** [comment]
As you can see
`tf.keras.metrics.PrecisionAtRecall(recall, num_thresholds=200, name=None, dtype=None)`
has 3 arguments
The argument here to notice is `num_thresholds` which is optional and Defaults to 200. The number of thresholds to use for matching the given recall.

Here's how precision is calculated:
1. Calculate recall at all the thresholds (200 thresholds by default).
2. Find the index of the threshold where the recall is closest to the requested value.
3. Compute precision at that index.

For more info you can refer to the [source code](https://github.com/tensorflow/tensorflow/blob/e5bf8de410005de06a7ff5393fafdf832ef1d4ad/tensorflow/python/keras/metrics.py#L1626).

Threshold : A float value or a python list/tuple of float threshold values in [0, 1]. A threshold is compared with prediction values to determine the truth value of predictions (i.e., above the threshold is `true`, below is `false`). One metric value is generated for each threshold value.

In setting Recall value in this case `tf.keras.metrics.PrecisionAtRecall` will consider recall value over all the classes not a specific class.

You need to write your own function if you want to calculate recall for a specific class or use binary classification where you have 2 class - the class you are interested in setting the recall value and rest of the classes binned as a single class.

**Comment by gowthamkpr** [comment]
@FrancescaAlf Did the above comment help you in resolving your problem?

**Comment by FrancescaAlf** [comment]
yes, thanks :)

**Comment by faustomorales** [comment]
In case anyone else stumbles upon this, I adapted the existing metrics to work in a multiclass setting using a subclass. Essentially, I just transform my `y_true` and `y_pred` into the binary equivalent before passing them.

```python
class RecallAtPrecisionByClass(tf.keras.metrics.RecallAtPrecision):
    def __init__(self, class_id, *args, **kwargs):
        super().__init__(*args, **kwargs)
        self.class_id = class_id

    def update_state(self, y_true, y_pred, sample_weight=None):
        y_true = tf.equal(y_true[:, self.class_id], 1)
        y_pred = tf.where(
            tf.equal(tf.argmax(y_pred, axis=-1), self.class_id),
            y_pred[:, self.class_id],
            tf.zeros_like(y_pred[:, self.class_id]),
        )
        return super().update_state(
            y_true=y_true, y_pred=y_pred, sample_weight=sample_weight
        )
```

Here's a quick example of this metric on some dummy data.

```python
y_true = np.array([
    [1, 0],
    [0, 1],
    [0, 1],
    [0, 1]
])
y_pred = np.array([
    [0.7, 0.3],
    [0.7, 0.3],
    [0.7, 0.3],
    [0.7, 0.3]
])

# Precision is only 25% but our minimum is 33%.
metric = RecallAtPrecisionByClass(class_id=0, precision=0.3333)
metric.update_state(y_true=y_true, y_pred=y_pred)
metric.result().numpy() # 0

# Here we exclude the final prediction so that the precision is 33%,
# which meets our minimum.
metric = RecallAtPrecisionByClass(class_id=0, precision=0.3333)
metric.update_state(y_true=y_true[:-1], y_pred=y_pred[:-1])
metric.result().numpy() # 1.0
    """
```
